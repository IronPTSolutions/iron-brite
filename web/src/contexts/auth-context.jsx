import { useContext, createContext, useState } from 'react';

const AuthContext = createContext();

export function AuthProvider({ children }) {
  const [user, setUser] = useState(localStorage.getItem('user') ? JSON.parse(localStorage.getItem('user')) : undefined);

  function login(user) {
    localStorage.setItem('user', JSON.stringify(user));
    setUser(user);
  }

  function logout() {
    localStorage.removeItem('user');
    setUser(undefined);
  }

  const contextData = {
    user,
    login,
    logout
  }
  return <AuthContext.Provider value={contextData}>{children}</AuthContext.Provider>
}

export function useAuthContext() {
  return useContext(AuthContext);
}
