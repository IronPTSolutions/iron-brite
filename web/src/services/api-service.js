import axios from 'axios';

const http = axios.create({
  baseURL: 'http://localhost:3000/api/v1/',
  withCredentials: true
})

http.interceptors.response.use(
  (response) => response.data,
  (error) => {
    if (error.response?.status === 401 &&
       window.location.pathname !== '/login') {
        localStorage.removeItem('user');
        window.location.assign('/login');
    } else {
      return Promise.reject(error);
    }
  }
)

const login = (user) => http.post('/sessions', user)

const listEvents = ({ city, limit, page }) => {

  limit = Number.isNaN(Number(limit)) || Number(limit) <= 0 ? 1 : limit;
  page = Number.isNaN(Number(page)) || Number(page) <= 0 ? undefined : page;

  return http.get('/events', { params: { city, limit, page }})
}

const getEvent = (id) => http.get(`/events/${id}`);

const deleteEvent = (id) => http.delete(`/events/${id}`);

export {
  login,
  listEvents,
  getEvent,
  deleteEvent
}