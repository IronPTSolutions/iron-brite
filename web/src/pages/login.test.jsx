import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import LoginPage from "./login";
import * as authContext from "../contexts/auth-context";
import * as apiService from "../services/api-service";
import { BrowserRouter } from "react-router-dom";

describe("<Login/>", () => {
  beforeEach(() => {
    cleanup();
  });

  test("happy case: render OK", () => {
    // Given
    vi.spyOn(authContext, "useAuthContext").mockReturnValue({
      login: vi.fn(),
    });

    // When
    render(<LoginPage />, { wrapper: BrowserRouter });

    // Then
    expect(screen.getByTestId("email-input")).toBeTruthy();

    expect(document.body).toMatchSnapshot();
  });

  test("happy case", async () => {
    // Given
    const login = vi.fn();

    vi.spyOn(authContext, "useAuthContext").mockReturnValue({
      login,
    });

    const loginServiceMock = vi.spyOn(apiService, "login").mockResolvedValue({
      response: {},
    });

    // When
    render(<LoginPage />, { wrapper: BrowserRouter });

    const email = screen.getByTestId("email-input");
    const password = screen.getByTestId("password-input");

    fireEvent.change(email, { target: { value: "john@example.com" } });
    fireEvent.change(password, { target: { value: "123" } });

    const form = screen.getByTestId("login-form");

    fireEvent.submit(form);

    // Then
    await waitFor(() => {
      expect(loginServiceMock).toHaveBeenCalledWith({
        email: "john@example.com",
        password: "123",
      });

      expect(login).toBeCalledWith({ response: {} });
    });
  });
});
