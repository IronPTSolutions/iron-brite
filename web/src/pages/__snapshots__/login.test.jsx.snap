// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Login/> > happy case: render OK 1`] = `
<body>
  <div>
    <div
      class="container py-4"
    >
      <div>
        <form
          data-testid="login-form"
        >
          <div
            class="input-group mb-1"
          >
            <span
              class="input-group-text"
            >
              <i
                class="fa fa-user fa-fw"
              />
            </span>
            <input
              class="form-control "
              data-testid="email-input"
              name="email"
              placeholder="user@example.org"
              type="email"
            />
          </div>
          <div
            class="input-group mb-2"
          >
            <span
              class="input-group-text"
            >
              <i
                class="fa fa-lock fa-fw"
              />
            </span>
            <input
              class="form-control  "
              data-testid="password-input"
              name="password"
              placeholder="****"
              type="password"
            />
          </div>
          <div
            class="d-grid"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
